docs/how-to/test-on-juju-openstack-provider.rst

```diff
@@ -0,0 +1,152 @@
+===============================
+Test on Juju OpenStack provider
+===============================
+
+This document explain how to setup the testing environment on top of the Juju OpenStack provider and how to run COU testing.
+
+
+Bootstrap juju controller on openstack cloud
+--------------------------------------------
+
+Follow the step on the `official document <https://juju.is/docs/juju/manage-controllers#heading--bootstrap-a-controller>`_ to bootstrap a Juju Openstack controller
+
+
+Deploy openstack with juju openstack-provider
+---------------------------------------------
+
+Follow the step on `STSStack Bundles <https://github.com/canonical/stsstack-bundles>`_
+to deploy the openstack cloud
+
+
+.. code::
+
+    git clone https://github.com/canonical/stsstack-bundles.git
+    cd openstack
+
+    # Some of the overlays may be missing here, check all the supported overlays with:
+    ./generate-bundle.sh --list-overlays
+
+    # ussuri-focal is the lowest version that COU supports
+    ./generate-bundle.sh --name some-juju-model-name -r ussuri -s focal --ovn --telemetry
+
+
+(Optional) sshuttle as proxy
+----------------------------
+
+If your openstack environment is behind the vpn and you have the bastion server, you can use sshuttle:
+
+.. code:: bash
+
+     sshuttle --ssh-cmd "ssh -i /home/myuser/.ssh/mykey" -v -r ubuntu@your-bastion-ip 10.5.0.0/16
+
+
+Install COU for testing
+-----------------------
+
+There are two ways to execute COU:
+
+- Use snap
+- Use Python
+
+Using python is more useful to debugging the code base bugs.
+However Snap is a official way to run COU on production, and we need to make sure all the parts in snap is working fine.
+
+
+(Optional) Prepare local python environment for COU
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One way to execute COU is to run it in your python environment
+
+.. code:: bash
+
+    virtualenv ./.venv
+    source .venv/bin/activate
+    pip install -e .
+
+    # Execute cou in python environment
+    python -m cou
+
+(Optional) Install COU from local snap build
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code:: bash
+   # Running snapcraft command in the project's root directory.
+   snapcraft
+
+   # Install local snap with --dangerous
+   sudo snap install ./LOCAL_SNAP_FILE --dangerous
+
+   # snap list command should show the COU
+   snap list
+
+(Optional) Copy the JUJU_DATA
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you don't want to mix the configuration with your local.
+
+.. code:: bash
+    export JUJU_DATA=./juju-data
+
+
+Execute the COU
+---------------
+
+Run cou plan and review the steps generate by COU
+
+.. code:: bash
+
+    python -m cou plan
```

Comment:
There are different. One run the cou from the code directly, one run the cou install in the python environment.

Reply:
It really depends on the user because you `cou plan` in python environment mean every time you update your code, you need to re-run `python install -e`.

`python -m cou plan` mean execute the module codebase directly so you don't have to re-install command.